src/v/archival/adjacent_segment_merger.h

```diff
@@ -71,6 +72,8 @@ class adjacent_segment_merger : public housekeeping_job {
     ss::abort_source _as;
     ss::gate _gate;
     ss::gate::holder _holder;
+    // Set to true if the topic is compacted
+    const bool _is_compacted{false};
```

Comment:
In `adjacent_segment_merger::run` we call `ntp_archiver::find_reupload_candidate`. That ends up calling `segment_collector::make_upload_candidate` with the `segment_collector_mode::collect_non_compacted` mode. There's some special handling in the collector for compacted segments which only kicks in when using the `collect_compacted` mode.

Reply:
The name `collect_non_compacted` is not very good. The `collect_compacted` signifies that the code is invoked from compacted reupload and it does all stuff that compacted reuploader needs. The `collect_non_compacted` means that the code is invoked from adjacent segment merger so it should just cut the region we asked it to cut and avoid doing any smarts.